### PR TITLE
Add Huc6280 Support

### PR DESCRIPTION
**Pull request by mre**
This adds support for the Hudson Soft / NEC HuC6280, the CPU powering the NEC `TurboGrafx-16` / PC Engine from 1987.

Apart from the additional instructions, it adds a 21-bit address space. Addressable memory starts at $2000, which is why I introduced support for relocating the zero-page. (Other variants are unaffected.)

The `T` (memory-operation) flag, which redirects ORA/AND/EOR/ADC and friends to operate on memory instead of the accumulator is not implemented yet. It's a bigger change, that will probably touch the other variants, so I left that out for now.

I found https://github.com/SingleStepTests/huc6280, which provides Mesen-generated tests. It's in JSON format, which is why I added a little translator to be able to run these tests. Found a few bugs this way. The corpus is quite large, which is why I didn't include it in the PR. We can consider changing that if worthwhile.

## TODO (for future PRs)

I checked what's missingand it's quite a lot, but it's probably okay to brush up this PR and merge what we already have. (Well, except for the mising Bus support, which I want to add before merging.)

- [ ] Sound (the 6-channel PSG), the programmable timer, and the 8-bit I/O port are not emulated (like all I/O they're left to the `Bus` implementation I think).
- [ ] The `SET` (T-flag) prefix instruction is not implemented, so the T-flag memory-operand mode for arithmetic/logic ops is unsupported.
- [ ] `CSL`/`CSH` (clock speed select) are no-ops; the emulator tracks per-instruction cycle counts, not wall-clock timing, so the 1.79/7.16 MHz switch has no effect. I *think* that's fine. Don't think that clock cycles should change when the clock speed gets changed.
- [ ] `ST0`/`ST1`/`ST2` just write the immediate to fixed memory offsets ($0000/$0002/$0003) and rely on a `Bus` implementation to intercept them as VDC port writes; there is no actual [VDC (Video Display Controller)](https://en.wikipedia.org/wiki/Hudson_Soft_HuC6270).
- [ ] The CPU core only ever emits 16-bit logical addresses; MMU logical-to-physical translation (MPR0-MPR7) is provided as a helper but the actual bank mapping must be done by the `Bus` (the integration tests use a translating `PhysBus` for this, but maybe we can remove that).
- [ ] Block-transfer instructions block until completion in a single step; their per-byte cost (6 cycles/byte) is added manually at execution time rather than coming from the cycle table, and interrupt behavior mid-transfer is not modelled. I am absolutely uncertain what to do with this one. It feels weird to manually add the cycle costs. Will have to think about that.
- [ ] The HuC6280-only opcodes share the global `base_cycles` table and execute match arms with all other variants; they're gated only by per-variant `decode`, so nothing structurally prevents another variant from reaching them if its decoder were changed to emit those instructions, but I think that's okay since adding another level of abstraction is just adding complexity for little gain.
- [ ] The `MPR` registers exist on every variant's `Registers` struct (zeroed and unused for non-HuC6280 parts), so the type doesn't distinguish HuC6280 state from the rest. We discussed moving more of that functionality into the Bus. Not sure if we can get rid of those registers entirely, though. I was considering to move the register definition into the variant trait, but it would add complexity and the number of unused registers per variant is pretty limited anyway.
- [ ] Only opcodes covered by the SingleStepTests suite are validated; untested edge cases (e.g. flag effects on rarely-used combinations) may differ from hardware. I am planning to do a larger test with some actual test roms (TurboGrafx-16 games) when I get the chance.

### CODE REVIEW

File: src/registers.rs [hunk 1]
```diff
@@ -197,6 +224,7 @@ impl Registers {
             stack_pointer: StackPointer(0), // Real hardware: random value on power-on
             program_counter: 0,             // Set by reset vector in practice
             status: Status::default(),
+            mpr: [0; 8], // HuC6280 MMU registers; unused by other variants
```
Comment (omarandlorraine) [inline src/registers.rs:245]: Is there a way to move this into the Bus trait somehow? That would be cleaner. But then, I have no idea how this works or what it needs to do.

File: src/registers.rs [hunk 2]
```diff
@@ -177,6 +177,33 @@ pub struct Registers {
...
+    /// produced by the `HuC6280` MMU, using the current mapping registers.
+    ///
+    /// The top three bits of `logical` select the mapping register (MPR0-MPR7);
+    /// the selected register supplies bits 13-20 of the physical address while
+    /// the low 13 bits pass through unchanged.
+    ///
+    /// For every non-HuC6280 variant the mapping registers are zero, so this is
+    /// the identity mapping within the low 64 KB.
+    #[must_use]
+    pub const fn physical_address(&self, logical: u16) -> u32 {
```
Comment (omarandlorraine) [inline src/registers.rs:203]: Question: Can the MMU functionality move into an implementation of the Bus trait? (that would include this function, `physical_address`, the `mpr` registers and the handling of relevant opcodes.

File: src/lib.rs [hunk 3]
```diff
@@ -141,6 +150,34 @@ pub trait Variant {
         0 // Default: no penalty
     }
 
+    /// Base address of the zero page for this variant.
+    ///
+    /// Standard 6502 family parts place the zero page at `$0000`. The `HuC6280`
+    /// relocates it to `$2000` (reached through mapping register MPR1), so all
+    /// zero-page and zero-page-indirect accesses are offset by this base.
+    #[must_use]
+    fn zero_page_base() -> u16 {
```
Comment (omarandlorraine) [inline src/lib.rs]: Very good. Other variants let you move the zeropage as well, so this is a welcome addition.

File: .gitignore [hunk 4]
```diff
@@ -65,4 +65,8 @@ TAGS.vi
 src/.DS_Store
 tmp.*.rs
 .vscode
-Cargo.lock
\ No newline at end of file
+Cargo.lock
```
Comment (omarandlorraine) [inline .gitignore:68]: Caution, Cargo.lock has snuck into .gitignore, consider if this is intentional

File: src/lib.rs (hunk 3 above)
Comment (omarandlorraine) [reply src/lib.rs]: But it should have a `&self` parameter since some variants like the 65816 call it the direct page and can move it around the whole address space

File: src/lib.rs [hunk 5]
```diff
@@ -141,6 +150,34 @@ pub trait Variant {
...
+        0x0000
+    }
+
+    /// Base address of the hardware stack page for this variant.
+    ///
+    /// Standard 6502 family parts place the stack at `$0100`. The `HuC6280`
+    /// relocates it to `$2100` (reached through mapping register MPR1).
+    #[must_use]
+    fn stack_base() -> u16 {
```
Comment (omarandlorraine) [inline src/lib.rs]: Question: should this have a `&self` parameter? (either just for symmetry to `zero_page_base`, or in case any Variant will need it)

**Review by omarandlorraine — APPROVED** [review]

File: src/registers.rs (hunk 1 above)
Comment (mre) [reply src/registers.rs:245]: Hmm... maybe? But then again, I believe that they are actual registers. See [here](https://jsgroth.dev/blog/posts/pc-engine-cpu/#memory-management):

> Software running on the HuC6280 operates on 16-bit memory addresses, same as 6502 software, but the HuC6280 has a builtin [MMU](https://en.wikipedia.org/wiki/Memory_management_unit) that expands the physical address space from 16-bit to 21-bit (2 MB address range). It’s extremely simple as far as MMUs go: it splits the 16-bit logical address space into eight 8 KB pages, and each page has its own 8-bit MPR (Memory Page Register) that maps it directly to an 8 KB physical page. It’s very similar to the memory-banking mappers seen in many NES / Game Boy / Master System / Game Gear cartridges, only it’s built into the CPU, so game cartridges don’t need to include their own mapper hardware (though [one game](https://en.wikipedia.org/wiki/Street_Fighter_II) still does due to its massive ROM size).

File: src/registers.rs (hunk 2 above)
Comment (mre) [reply src/registers.rs:203]: Thx, I can look into that.

File: src/lib.rs (hunk 3 above)
Comment (mre) [reply src/lib.rs]: Good idea. Could also be `const` I guess.

File: .gitignore (hunk 4 above)
Comment (mre) [reply .gitignore:68]: Ah, this is just an extra newline that got added. 😉 (See comment in the diff: "No newline at end of file")

File: src/lib.rs (hunk 3 above)
Comment (mre) [reply src/lib.rs]: Oh, it can't be `const` if it's also `&self`, I guess. 🤔

File: src/lib.rs (hunk 5 above)
Comment (mre) [reply src/lib.rs]: Yeah, probably.

File: src/lib.rs (hunk 3 above)
Comment (mre) [reply src/lib.rs]: Ah yeah, const + &self does work. https://play.rust-lang.org/?version=stable&mode=debug&edition=2024&gist=277425ef9ade0c242ae6e166fda5bb4f

File: src/registers.rs (hunk 2 above)
Comment (mre) [reply src/registers.rs:203]: Right now we have to shadow the MPR registers because translation runs inside `bus.get_byte()`, which isn't idea. The MPRs live in the cpu, so the bus needs to copy. What's worse is that we have to create a copy whenever `TAM` runs...

I don't think there's an easy solution here.

We could try moving MPRs into the bus (then the bus needs MMU-specific trait methods + HuC6280 can no longer run on a plain `Memory` bus), or we try moving translation into the CPU (then `Bus` has to take 21-bit physical addresses, which pollutes the generic core for one variant).

Both of those make the generic core worse to fix one variant. So I'd say the current tradeoff (generic u16 `Bus`, MMU as this thin integrator-side adapter) is the more idiomatic way.

But there are still ways to make it more ergonomic. I noticed that there's quite a bit of redundancy, and I'll push a commit in a sec to fix that.

Also voting to add `set_mapping_register` to the `Bus` trait. The default implementation does nothing, which is correct for all variants without an MMU. I will add this method, but we can remove it again if it doesn't carry its own weight.

File: src/registers.rs (hunk 1 above)
Comment (mre) [reply src/registers.rs:245]: I would keep it in there for now.
